### ISSUE TITLE
Unclear docs and error message for --match-conditions format

### ISSUE THREAD

**Issue by ssolders** [Azure/azure-cli#29010]
**Related command**

```
az network application-gateway waf-policy custom-rule create \
--resource-group my-group \
--policy-name my-policy \
--name azuredevopsip \
--priority 10 \
--rule-type MatchRule \
--match-conditions '[{"variables": "RemoteAddr","operator":"IPMatch","values": ["'"$CURRENT_IP"'"]}]' \
--action Allow
```

**Is your feature request related to a problem? Please describe.**
When reading docs / looking at examples, I was not able to find any relevant information about how `--match-conditions` should be formatted.

When I ran this command I got the error `Failed to parse '--match-conditions' argument: Model 'AAZObjectArg' has no field named 'RemoteAddr'`

When running `--match-conditions` with `??` for docs it suggested `variables [Required] Space-separated list of variables to use when matching. Variable values: RemoteAddr, RequestMethod, QueryString, PostArgs, RequestUri, RequestHeaders, RequestBody, RequestCookies. Try "??" to show more.`.

Documentation on [azure-cli docs](https://learn.microsoft.com/en-us/cli/azure/network/application-gateway/waf-policy/custom-rule?view=azure-cli-latest) doesn't present any examples on  `--match-conditions`.

Solution to this was to pass variables as an array of objects, where each key should have one of the values from `RemoteAddr, RequestMethod, QueryString, PostArgs, RequestUri, RequestHeaders, RequestBody, RequestCookie`.

Correct format:
```
--match-conditions '[{"variables":[{"variableName":"RemoteAddr"}],"operator":"IPMatch","values": ["'"$CURRENT_IP"'"]}]'
```

**Describe the solution you'd like**
The error message returned did not describe clear enough what the problem was, it suggested the variables key was not correct, but the docs referred to that it should be one of the `RemoteAddr, RequestMethod, QueryString, PostArgs, RequestUri, RequestHeaders, RequestBody, RequestCookie`.

**Describe alternatives you've considered**
Including example payloads of --match-conditions.

**Additional context**

**Comment by yonzhan** [comment]
Thank you for opening this issue, we will look into it.

**Comment by necusjz** [comment]
<img width="557" alt="image" src="https://github.com/Azure/azure-cli/assets/12371639/4ae7149c-9103-4cca-baa8-86212df8dc3d">

documentation of shorthand syntax https://github.com/Azure/azure-cli/blob/dev/doc/shorthand_syntax.md

**Comment by ssolders** [comment]
> <img alt="image" width="557" src="https://private-user-images.githubusercontent.com/12371639/333498957-4ae7149c-9103-4cca-baa8-86212df8dc3d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1Mzc0MDEsIm5iZiI6MTcxNjUzNzEwMSwicGF0aCI6Ii8xMjM3MTYzOS8zMzM0OTg5NTctNGFlNzE0OWMtOTEwMy00Y2NhLWJhYTgtODYyMTJkZjhkYzNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI0VDA3NTE0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRjMDQ0OTgxNDdlYzE0NWI0NzJiMjIxMzgxMTcwNTkyMmM2ZmZiNmMxOTRlOGVlNzY2MmNlMTk5OWFiNDQ3MjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZowQq3KhKwbWvjsPHZbIahCbPIwyZazIHM8m14wLQjM">
> documentation of shorthand syntax https://github.com/Azure/azure-cli/blob/dev/doc/shorthand_syntax.md

Yea I found my way there, the error message I got linked to that page, but I didn't get any help from it.
If the docs had also contained a few examples it would have helped a lot, a few common use cases.

**Comment by pmeems** [comment]
I agree with @ssolders it is hard to understand the correct syntax.
We need improved documentation with examples how to use it.

I've tried several variants of this in my PowerShell script, but can't get it to work:

```
az network application-gateway waf-policy custom-rule create `
    --resource-group $ResourceGroupName `
    --name $RuleName `
    --policy-name $WafPolName `
    --action $Action `
    --priority $Priority `
    --match-conditions '[{variables:[{variableName:'$matchVariable'}],operator:'$Operator',values: '$value'}]' `
    --rule-type MatchRule
```
For the above command I get this error:

```
Error formatting a string: Input string was not in a correct format. Failure to parse near offset 4. Unexpected closing brace without a corresponding opening brace..
```

**Comment by pmeems** [comment]
After a full day of trial-and-error, Googling, and even a little bit of ChatGPT I found a working solution:

```
$matchCondition = @"
[{
    "variables": [
        {
            "variableName": "$matchVariable"
        }
    ],
    "operator": "$Operator",
    "values": [
        "$value"
    ]
}]
"@

az network application-gateway waf-policy custom-rule create `
    --resource-group $ResourceGroupName `
    --name $RuleName `
    --policy-name $WafPolName `
    --action $Action `
    --priority $Priority `
    --match-conditions $matchCondition `
    --rule-type MatchRule
```

I hope this will help somebody.

**Comment by edsaantos** [comment]
Com base na resposta do @pmeems consegui executar no bash dessa forma:

```
az network application-gateway waf-policy custom-rule create --action Allow \
--name $rule_name --policy-name $policy --priority $priority \
--resource-group $rg --rule-type MatchRule \
--match-conditions "[{"variables":[{"variableName":"RemoteAddr"}],"operator":"IPMatch","values":["$agent_ip"]}]"
```

**Comment by ilyaro** [comment]
Hi

IMHO, it just requires a change of manual from _list_ type to **dict** type

$ az network application-gateway waf-policy custom-rule create   --policy-name AllowAWSBlockAllWAFPolicy   --resource-group tf   --name AllowSpecificIPs   --priority 1   --action Allow   --rule-type MatchRule   --match-conditions "[{"variables":["RemoteAddr"],"operator":"IPMatch","values":["3.3.2.3"]}]"


Failed to parse '--match-conditions' argument: Invalid at [0] : Invalid 'variables' : Invalid at [0] : **dict type value expected**, got 'RemoteAddr'(<class 'str'>)

It is written in error that the dictionary type is expected when I gave it the list as written in the manual

In the latest version of az cli
$ az version
  "azure-cli": "2.67.0",

In the documentation, it is written that "variables" is a list type

$ az network application-gateway waf-policy custom-rule create --match-conditions ?? 
variables [Required] : Space-separated **list** of variables to use when matching. Variable values:
                          RemoteAddr, RequestMethod, QueryString, PostArgs, RequestUri,

So the fix here IMHO is to change the documentation: 
variables [Required] : **dict** type of variables and give example: --match-conditions "[{"variables":[{"variableName":"RemoteAddr"}],"operator":"IPMatch","values":["3.3.2.3"]}]"

**Comment by cg-techgw** [comment]
Agree, the only useful documentation (including the suggested "??" option) was @pmeems comment on this thread. I too tried every google I could think, and resorted to Google Gemini (which interestingly got it the exact same kind of wrong that I did), none of which worked barring the bizarre solution of editing the json to use **different keys than the response from the same API** (for example using the related command "waf-policy show"). This is clearly not sensible, especially when the documentation is unhelpful.
